src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +237,814 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+        try:
+            ref1_accesses = self._compute_accesses(ref1, preceding_t1, task1)
+            ref2_accesses = self._compute_accesses(ref2, preceding_t2, task2)
+        except GenerationError:
+            # If we get a Generation error from compute_accesses, then we found
+            # an access that isn't able to be handled by PSyclone, so
+            # dependencies based on it need to be handled by a taskwait
+            return False
+
+        # If the first access in each accesses contains a Reference
+        # we should check that both are to the same symbol
+        ref1_ref = ref1_accesses[0].walk(Reference)
+        ref2_ref = ref2_accesses[0].walk(Reference)
+        if len(ref1_ref) > 0 and len(ref2_ref) == 0:
+            # Found a pair of dependencies on the same array which are not
+            # valid under OpenMP, as one contains a Reference while the
+            # other does not.
+            return False
+        if len(ref1_ref) == 0 and len(ref2_ref) > 0:
+            # Found a pair of dependencies on the same array which are not
+            # valid under OpenMP, as one contains a Reference while the
+            # other does not.
+            return False
+        # If we have elements in the accesses lists, check the first element is
+        # the same in both.
+        if len(ref1_ref) > 0 and ref1_ref[0] != ref2_ref[0]:
+            # Found a pair of dependencies on the same array which are not
+            # supported in PSyclone, as they are both References but to
+            # different variables.
+            return False
+
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.backend.sympy_writer import SymPyWriter
+        # Now we know if there is a Reference, both are to the same Reference
+        if len(ref1_ref) > 0:
+            # Handle reference case
+            values = []
+            r1_min = sys.maxsize
+            r1_max = 0
+            for member in ref1_accesses:
+                # If its a reference only we can ignore it as we know both
+                # feature the Reference.
+                if isinstance(member, Reference):
+                    # If we find the Reference then we need to set the min
+                    # to 0 as we have Reference+0
+                    r1_min = 0
+                    continue
+                # We know its a BinaryOperation of the Reference and something.
+                # Take the second child of the BinaryOperation to compute
+                values.append(member.children[1])
+            sympy_ref1s = SymPyWriter.convert_to_sympy_expressions(values)
+
+            values = []
+            for member in sympy_ref1s:
+                val = int(member)
+                values.append(val)
+                if val < r1_min:
+                    r1_min = val
+                if val > r1_max:
+                    r1_max = val
+
+            val2s = []
+            for member in ref2_accesses:
```

Comment:
Its slightly different - here we do ignore the Reference object. I'm not sure its worth saving the lines here since we'd have to pull out all non-Reference members in each accesses list first.